### ISSUE TITLE
Change `stimes @Void n x` to be `x` 

### ISSUE THREAD

**Issue by phadej** [haskell/core-libraries-committee#95]
Currently

```haskell
Prelude Data.Semigroup Data.Void> stimes @Void 1 (error "VOID")
*** Exception: VOID
CallStack (from HasCallStack):
  error, called at <interactive>:6:17 in interactive:Ghci3
Prelude Data.Semigroup Data.Void> stimes @Void (-1) (error "VOID")
*** Exception: stimesIdempotent: positive multiplier expected
```

however

```haskell
Prelude Data.Semigroup Data.Void> stimes  @() (-1) ()
()
Prelude Data.Semigroup Data.Void> stimes  @() 1 ()
()
```

The motivation is to allow refactoring `base` so `Void` can be defined earlier. That's probably a reason why `stimes` for `()` is defined as it is. An alternative is to change `stimes @()` to also fail for non-positive integral values.

### Diff

```diff
 instance Semigroup Void where
     a <> _ = a
-    stimes = stimesIdempotent
+    stimes _ x = x
```

I don't expect this change to affect anyone.

The class says

> Given that this works on a [Semigroup](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Semigroup.html#t:Semigroup) it is allowed to fail if you request 0 or fewer repetitions, and the default definition will do so.

So it's not required to fail if `n <= 0`, so it could as well return an absurd `Void` value. We change one bottom to another, should not be a big deal.

### historical note

In very old `void` its instance was defined as, e.g. https://hackage.haskell.org/package/void-0.5.12/docs/src/Data-Void.html#Void

```haskell
instance Semigroup Void where
  a <> _ = a
  times1p _ a = a
```

i.e. not using `stimesIdempotent`. I guess it was changed when migrating to base. I don't know why.

**Comment by phadej** [comment]
MR https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9112

**Comment by monoidal** [comment]
Could you explain the motivation more? In the MR the module `GHC.Base` already imports `stimesIdempotentMonoid` and adding an additional import of the simpler function `stimesIdempotent` shouldn't change much. Or am I missing something?

Looks like the default was changed in https://github.com/ekmett/semigroups/commit/bb03b9365a5708ec382f7d8ec3f37000e201fa67. I guess it was for consistency with other semigroup instances (like `Last a` or `Min a`), where it makes more sense to fail for n <= 0 than to return `x`.

>  An alternative is to change `stimes @()` to also fail for non-positive integral values.

`()` forms a group (like `Sum a`) so it makes sense to support negative exponents. For `Void`, there's no neutral element and this reasoning is shakier.

This proposal is just trading one `_|_` for another `_|_` - I'm not objecting, just poking at the design space.

**Comment by Bodigrim** [comment]
I'm mildly in favor: trading one bottom for another makes the change non-breaking (anything which was not broken before) and `stimes = const id` should generate slightly less code than `stimesIdempotent` with two branches.

**Comment by ekmett** [comment]
no objection here

**Comment by Bodigrim** [comment]
Dear CLC members, let's vote on the proposal to change `stimes @Void n x` to be simply `x` instead of calling [`stimesIdempotent`](https://hackage.haskell.org/package/base-4.17.0.0/docs/src/Data.Semigroup.Internal.html#stimesIdempotent) (which throws an error when `n <= 0`). @tomjaguarpaw @cgibbard @chessai @emilypi @mixphix 

----

+1 from me.

**Comment by emilypi** [comment]
+1 easy

**Comment by chessai** [comment]
+1

**Comment by tomjaguarpaw** [comment]
+1

**Comment by mixphix** [comment]
+1

**Comment by Bodigrim** [comment]
With 5 votes in favour, the proposal is approved. Thanks all.

**Comment by chshersh** [comment]
I'm trying to summarise the state of this proposal as part of my volunteering effort to track the progress of all `approved` CLC proposals.

| Field | Value |
|---|---|
| **Authors**            |  @phadej |
| **Status**             |  merged  |
| **`base` version**     | Unknown  |
| **Merge Request (MR)** | https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9112 |
| **Blocked by**         | nothing |
| **CHANGELOG entry**    | missing (please, [raise an MR to `base` to update changelog][changelog])  |
| **Migration guide**    |  not needed |

[changelog]: https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/changelog.md

Please, let me know if you find any mistakes 🙂  

<hr>

@phadej I don't see a changelog entry for this code change, and thus it's difficult to see, which `base` version implements it. Could you raise a PR to update `base` changelog?

**Comment by Bodigrim** [comment]
It's been released with `base-4.18`. A changelog entry would be most welcome.
